Add third-party JS loading post (#1221)
GoogleChrome/web.dev: src/site/content/en/fast/efficiently-load-third-party-javascript/index.md

---
layout: post
title: Efficiently load third-party JavaScript

> subhead: Avoid common pitfalls of using third-party scripts and improve load times and user experience.

Reviewer: ```suggestion
subhead: Avoid common the pitfalls of using third-party scripts to improve load times and user experience.
```

authors:
  - mihajlija
date: 2019-07-29
description: |

>   Learn how to avoid common pitfalls of using third-party scripts and improve load times and user experience.

Reviewer: ```suggestion
  Learn how to avoid the common pitfalls of using third-party scripts to improve load times and user experience.
```

hero: hero.jpg
alt: Aerial view of shipping containers.  
tags:

>   - post # post is a required tag for the article to show up in the blog.

Reviewer: ```suggestion
```
Given the third-party posts are kind of interdependent, I'm thinking maybe these shouldn't go on the blog. What do you think @robdodson?

  - performance
  - third-party
  - javascript
---

If a third-party script is [slowing down](/third-party-javascript/) your page load, you have two options to improve performance:

* Remove it if it doesn't add clear value to your site.

* Optimize the loading process.

> This guide explains how to optimize the loading process of third-party scripts with the following techniques: 

Reviewer: ```suggestion
This post explains how to optimize the loading process of third-party scripts with the following techniques: 
```
web.dev goes with "post" rather than "guide." I'm adding it to our word list :)

1. Using the `async` or `defer` attribute on `<script>` tags

2. Establishing early connections to required origins

3. Lazy-loading

4. Optimizing how you serve third-party scripts 

## Use `async` or `defer`

Because [synchronous scripts](link-to-intro) delay DOM construction and rendering, you should always load third-party scripts asynchronously unless the script has to run before the page can be rendered.

> The `async` and `defer` attributes tell the browser that it may go on parsing the HTML while loading the script in the background, and then execute the script after it loads. This way, script downloads don’t block DOM construction and page rendering. The result is that the user can see the page before all scripts have finished loading.

Reviewer: ```suggestion
The `async` and `defer` attributes tell the browser that it may go on parsing the HTML while loading the script in the background, and then execute the script after it loads. This way, script downloads don't block DOM construction and page rendering. The result is that the user can see the page before all scripts have finished loading.
```
Curly apostrophe to straight.

```html
<script async src="script.js">

<script defer src="script.js">
```

The difference between `async` and `defer` is when they start executing the scripts.

### `async`

> Scripts with the `async` attribute execute at the first opportunity after they finish downloading and before the window’s [load](https://developer.mozilla.org/en-US/docs/Web/Events/load) event. This means it’s possible (and likely) that `async` scripts will not be executed in the order in which they appear in the HTML. It also means they can interrupt DOM building if they finish downloading while the parser is still at work.

Reviewer: ```suggestion
Scripts with the `async` attribute execute at the first opportunity after they finish downloading and before the window's [load](https://developer.mozilla.org/en-US/docs/Web/Events/load) event. This means it's possible (and likely) that `async` scripts will not be executed in the order in which they appear in the HTML. It also means they can interrupt DOM building if they finish downloading while the parser is still at work.
```
Curly apostrophe to straight.

![Diagram of parser blocking script with async attribute](async@2x.png)

### `defer`

Scripts with the `defer` attribute execute after HTML parsing is completely finished, but before the [`DOMContentLoaded`](https://developer.mozilla.org/en-US/docs/Web/Events/DOMContentLoaded) event. `defer` guarantees scripts will be executed in the order they appear in the HTML and will not block the parser.

![Diagram of parser flow with a script with defer attribute](defer@2x.png)

* Use `async` if it's important to have the script run earlier in the loading process. 

* Use `defer` for less critical resources. A video player that's below-the-fold, for example.

> Using these attributes can significantly speed up page load. For example, [Telegraph recently deferred all of their scripts](https://medium.com/p/a0a1000be5#4123), including ads and analytics, and improved the First Ad Loaded metric by an average of four seconds.

Reviewer: Is First Ad Loaded a metric that Telegraph devised? (I didn't find anything other than the Telegraph Medium post with a quick Google search.) If so, it seems weird to present it as proper noun?

{% Aside %}

> Analytics scripts are usually loaded early to not miss any valuable analytics data. Fortunately, there are [patterns to initialize analytics lazily](https://philipwalton.com/articles/the-google-analytics-setup-i-use-on-every-site-i-build/) while retaining early page-load data.

Reviewer: ```suggestion
Analytics scripts are usually loaded early so you don't miss any valuable analytics data. Fortunately, there are [patterns to initialize analytics lazily](https://philipwalton.com/articles/the-google-analytics-setup-i-use-on-every-site-i-build/) while retaining early page-load data.
```
Just a suggestion. I stumbled over the syntax on the first read.

{% endAside %}

## Establish early connections to required origins 

> Establishing connections takes significant time on slow networks. This is especially true for secure connections, which may involve DNS lookups, redirects, and several round trips to the final server that handles the user’s request. 

Reviewer: ```suggestion
Establishing connections takes significant time on slow networks. This is especially true for secure connections, which may involve DNS lookups, redirects, and several round trips to the final server that handles the user's request. 
```
Curly apostrophe to straight.

You can save 100–500 ms by establishing early connections to important third-party origins.

Two [`<link>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/link) types can help here:

* `preconnect`

* `dns-prefetch`

### `preconnect`

> `<link rel="preconnect">` informs the browser that your page intends to establish a connection to another origin, and that you’d like the process to start as soon as possible. When the request for a resource from the preconnected origin is made, the download starts immediately. 

Reviewer: ```suggestion
`<link rel="preconnect">` informs the browser that your page intends to establish a connection to another origin, and that you'd like the process to start as soon as possible. When the request for a resource from the preconnected origin is made, the download starts immediately. 
```
Curly apostrophe to straight.

```html
<link rel="preconnect" href="https://cdn.example.com">
```

{% Aside 'caution' %}

> 

Reviewer: ```suggestion
```

> Only preconnect to critical origins you will use soon because if the connection isn’t used within 10 seconds the browser closes it. There’s a limit to the number of simultaneous connections a browser can handle (for example, Chrome can handle six), and unnecessary preconnecting can delay other important resources.

Reviewer: ```suggestion
Only preconnect to critical origins you will use soon; if the connection isn't used within 10 seconds, the browser closes it. There's a limit to the number of simultaneous connections a browser can handle. (For example, Chrome can handle six.) Unnecessary preconnecting can delay other important resources.
```
Suggestions to improve readability by breaking up the long sentences.

{% endAside %}

### `dns-prefetch`

`<link rel="dns-prefetch>` handles a small subset of what is handled by `<link rel="preconnect">`.  Establishing a connection involves the DNS lookup and TCP handshake, and for secure origins, TLS negotiations. `dns-prefetch` instructs the browser to only resolve the DNS of a specific domain before it has been explicitly called.

> Because there’s a limit to the number of connections a browser can open, for less critical third-party origins use `<link rel=dns-prefetch>`. 

Reviewer: ```suggestion
Because there's a limit to the number of connections a browser can open, for less critical third-party origins use `<link rel=dns-prefetch>`. 
```
Curly apostrophe to straight.

```html
<link rel="dns-prefetch" href="http://example.com">
```

> [Browser support for `dns-prefetch`](https://caniuse.com/#search=dns-prefetch) is slightly different from [`preconnect` support](https://caniuse.com/#search=preconnect), so `dns-prefetch` can serve as a fallback for browsers that don’t support `preconnect`. Use separate link tags to implement this safely: 

Reviewer: ```suggestion
[Browser support for `dns-prefetch`](https://caniuse.com/#search=dns-prefetch) is slightly different from [`preconnect` support](https://caniuse.com/#search=preconnect), so `dns-prefetch` can serve as a fallback for browsers that don't support `preconnect`. Use separate link tags to implement this safely: 
```
Curly apostrophe to straight.

```html
<link rel="preconnect" href="http://example.com">
<link rel="dns-prefetch" href="http://example.com">
```
## Lazy-load third-party resources

> Embedded third-party resources can be a big contributor to slow page speed when constructed poorly. If they aren’t critical or are below the fold (that is, if users have to scroll to view them), lazy-loading is a good way to improve page speed and paint metrics. This way users will get the main page content faster and have a better experience.

Reviewer: ```suggestion
Embedded third-party resources can be a big contributor to slow page speed when constructed poorly. If they aren't critical or are below the fold (that is, if users have to scroll to view them), lazy-loading is a good way to improve page speed and paint metrics. This way, users will get the main page content faster and have a better experience.
```

<figure class='w-figure w-figure--inline-left'>
<img src="lazy-load.png" alt="A diagram of a webpage shown on a mobile device with scrollable content extending beyond the screen. The content that's below-the-fold is desaturated because it's not loaded yet.">
</figure>

> One way to do that is to lazy-load third-party content after the main page content loads. Ads are a good candidate for this approach.

Reviewer: ```suggestion
One effective approach is to lazy-load third-party content after the main page content loads. Ads are a good candidate for this approach.
```
To avoid the ambiguous "that."

> Ads are an important source of income for many sites, but users come for the content. By lazy-loading ads and delivering the main content faster, you can increase the overall viewability percentage of an ad. For example, MediaVine switched to [lazy-loading ads](https://www.mediavine.com/lazy-loading-ads-mediavine-ads-load-200-faster/) and saw a 200% improvement in page load speed. DoubleClick have guidance on how to lazy-load ads in their [official documentation](https://support.google.com/dfp_premium/answer/4578089#lazyloading). 

Reviewer: ```suggestion
Ads are an important source of income for many sites, but users come for the content. By lazy-loading ads and delivering the main content faster, you can increase the overall viewability percentage of an ad. For example, MediaVine switched to [lazy-loading ads](https://www.mediavine.com/lazy-loading-ads-mediavine-ads-load-200-faster/) and saw a 200% improvement in page load speed. DoubleClick provides guidance on how to lazy-load ads in its [official documentation](https://support.google.com/dfp_premium/answer/4578089#lazyloading). 
```
In American English, collective nouns are singular when the members of the group are acting together.

An alternative approach is to load third-party content only when users scroll down to that section of the page.

> [Intersection Observer](https://developers.google.com/web/updates/2016/04/intersectionobserver) is a browser API that efficiently detects when an element enters or exits the browser's viewport and it can be used to implement this technique.

Reviewer: Should this and the following paragraph be attached to the preceding paragraph--to make it clearer that they're all about lazy-loading based on scroll position?

[lazysizes](https://web.dev/use-lazysizes-to-lazyload-images/) is a popular JavaScript library for lazy-loading images and [`iframes`](http://afarkas.github.io/lazysizes/#examples). It supports YouTube embeds and [widgets](https://github.com/aFarkas/lazysizes/tree/gh-pages/plugins/unveilhooks). It also has [optional support](https://github.com/aFarkas/lazysizes/blob/097a9878817dd17be3366633e555f3929a7eaaf1/src/lazysizes-intersection.js) for IntersectionObserver. 

{% Aside 'caution' %}

> Be careful when lazy-loading resources with JavaScript. If JavaScript fails to load, perhaps due to flaky network conditions, your resources won’t load at all. Keep an eye out for [native lazy-loading support for images and iframes](https://addyosmani.com/blog/lazy-loading/) that’s coming soon to Chrome.

Reviewer: ```suggestion
Be careful when lazy-loading resources with JavaScript. If JavaScript fails to load, perhaps due to flaky network conditions, your resources won't load at all. Keep an eye out for [native lazy-loading support for images and iframes](https://addyosmani.com/blog/lazy-loading/) that's coming soon to Chrome.
```
Curly apostrophe to straight.

{% endAside %}

## Optimize how you serve third-party scripts

> ### Use third-party CDN hosting

Reviewer: Sorry to keep coming back to this issue, but I want to make sure the advice we're providing is clear. Based on the sections that came before, this heading makes it seem like we're proposing CDNs as a viable option. But we're not really, at least as I read the post. If we think people should use self-hosting or CDNs with a service worker, I kinda think we should just come out and say that. @katejeffreys and @mihajlija, what do y'all think?

> It’s common for third-party vendors to provide URLs of JavaScript files that they host, usually on a [CDN](https://en.wikipedia.org/wiki/Content_delivery_network). The benefits of this approach are that you can get started quickly—just copy and paste the URL—and there’s no maintenance overhead. The third-party vendor handles server configuration and script updates.

Reviewer: ```suggestion
It's common for third-party vendors to provide URLs of JavaScript files that they host, usually on a [content delivery network (CDN)](https://en.wikipedia.org/wiki/Content_delivery_network). The benefits of this approach are that you can get started quickly—just copy and paste the URL—and there's no maintenance overhead. The third-party vendor handles server configuration and script updates.
```
It's web.dev convention to spell out an acronym on first use--to support less experienced devs.

> But because they are not on the same origin as the rest of your resources, loading files from a public CDN comes with a network cost. The browser needs to perform a DNS lookup, establish a new HTTP connection, and, on secure origins, perform an SSL handshake with the vendor’s server. 

Reviewer: ```suggestion
But because they are not on the same origin as the rest of your resources, loading files from a public CDN comes with a network cost. The browser needs to perform a DNS lookup, establish a new HTTP connection, and, on secure origins, perform an SSL handshake with the vendor's server. 
```
Curly apostrophe to straight.

> When you use files from third-party servers, you rarely have control over caching. Relying on someone else’s caching strategy might cause scripts to be unnecessarily re-fetched from the network too often.

Reviewer: ```suggestion
When you use files from third-party servers, you rarely have control over caching. Relying on someone else's caching strategy might cause scripts to be unnecessarily re-fetched from the network too often.
```
Curly apostrophe to straight.

{% Aside %}

> One possible benefit of serving third-party code this way is cross-domain caching. Sometimes a resource served from a public CDN may already be cached in the user’s browser if they previously visited another site that used the same resource from the same public CDN. This optimization[ doesn’t happen in Safari](https://andydavies.me/blog/2018/09/06/safari-caching-and-3rd-party-resources/) due to security reasons and it’s hard to tell how common it is in other browsers.

Reviewer: ```suggestion
One possible benefit of using a CDN is cross-domain caching. Sometimes a resource served from a public CDN may already be cached in the user's browser if they previously visited another site that used the same resource. However, this optimization[ doesn't happen in Safari](https://andydavies.me/blog/2018/09/06/safari-caching-and-3rd-party-resources/), and it's hard to tell how common it is in other browsers.
```
@katejeffreys @mihajlija: I trimmed this a little. Aside from that, as noted above, I feel like we're kinda talking out of both sides of our mouths. I think maybe this needs to be moved to the body copy with another sentence basically saying "Look, odds are one of the two strategies below is a better bet." What do you think?

{% endAside %}

### Self-host third-party scripts

> Self-hosting third-party scripts is an option that gives you more control over a script’s loading process. By self-hosting you can: 

Reviewer: ```suggestion
Self-hosting third-party scripts is an option that gives you more control over a script's loading process. By self-hosting you can: 
```
Curly apostrophes to straight.

* Reduce DNS lookup and round-trip times. 
* Improve [HTTP caching](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching) headers.
* Take advantage of [HTTP/2 server push](https://developers.google.com/web/fundamentals/performance/http2/). 

For example, Casper managed to [shave 1.7 seconds](https://medium.com/caspertechteam/we-shaved-1-7-seconds-off-casper-com-by-self-hosting-optimizely-2704bcbff8ec) off load time by self-hosting an A/B testing script.

> Self-hosting comes with one big downside though: scripts can go out of date and won’t get automatic updates when there’s an API change or a security fix. 

Reviewer: Can we note a way to mitigate this issue? Like, say, package management?

Author: cc @addyosmani #phoneafriend

> Self-hosting comes with one big downside though: scripts can go out of date and won’t get automatic updates when there’s an API change or a security fix. 

Reviewer: ```suggestion
Self-hosting comes with one big downside though: scripts can go out of date and won't get automatic updates when there's an API change or a security fix. 
```
Curly apostrophes to straight.
